src/client/renderingengine.cpp

```diff
@@ -203,6 +212,149 @@ void RenderingEngine::setXorgClassHint(
+			<< std::endl;
+		return;
+	}
+
+	verbosestream << "Setting Xorg _NET_WM_PID window property"
+		<< std::endl;
+
+	Window x11_win = (Window)video_data.OpenGLLinux.X11Window;
+
+	// Set the _NET_WM_PID for the window. 
+
+	Atom NET_WM_PID = XInternAtom(x11_dpl,
+		"_NET_WM_PID",
+		false);
+
+	pid_t pid = getpid();
+	infostream << "PID is '" << (long)pid << "'"
+		<< std::endl;
+
+	XChangeProperty(x11_dpl, x11_win, NET_WM_PID, 
+			XA_CARDINAL, 32, PropModeReplace, 
+			(unsigned char *) &pid,1);
+
+	// According to the Extended Window Manager Hints spec if _NET_WM_PID
+	// is set then WM_CLIENT_MACHINE must also be set.
+
+	verbosestream << "Attempting to determine Fully Qualified Domain Name for client"
+		<< std::endl;
+
+	// Get the X client hostname which will be used to get address info 
+	// further down. If a canonical name can't be determined then 
+	// hostname will be used for the FQDN. 
+	//
+	// There is no way to know for sure what the max length of the hostname
+	// will be as in modern linux distributions this is a configurable option. 
+	// Most likely it will be be no longer than 64 bytes as this is the default
+	// but it can be longer ... instead of trying to guess the length use
+	// the results of gethostname to determine the length. 
+
+	size_t hostname_len=128;
+	char* hostname=0;
+	while (1) {
+		// (Re)allocate buffer of length hostname_len.
+		char* realloc_hostname=(char *) realloc(hostname,hostname_len);
+		if (realloc_hostname==0) {
+			free(hostname);
+			// This should probably raise an error or exit the program 
+			// if this happens. 
+			errorstream << "Failed to allocate memory for hostname"
+				<< std::endl;
+			return;
+		}
+		hostname=realloc_hostname;
+
+		// Terminate the buffer.
+		hostname[hostname_len-1]=0;
+
+		// Offer all but the last byte of the buffer to gethostname.
+		if (gethostname(hostname,hostname_len-1)==0) {
+			size_t count=strlen(hostname);
+			if (count<hostname_len-2) {
+				// Break from loop if hostname definitely not truncated
+				break;
+			}
+		}
+
+		// Double size of buffer and try again.
+		hostname_len*=2;
+	}
+
+	// The EWMH spec specifies that WM_CLIENT_MACHINE should be set to the 
+	// Fully Qualified Domain Name (FQDN) of the client. Here, attempt 
+	// to determine the FQDN using getaddrinfo. If that fails then fall
+	// back to the hostname discovered above. 
+
+	struct addrinfo hints={0};
+	hints.ai_family=AF_UNSPEC;
+	hints.ai_flags=AI_CANONNAME;
+
+	// gethostname and getaddrinfo do not support IDN (thank the 
+	// flying spaghetti monster) this means that fqdn will be represented
+	// in ASCII (any IDN names would be converted using punycode).
+	std::string *fqdn_ptr;          
+
+	struct addrinfo* res=0;
+	if (getaddrinfo(hostname,0,&hints,&res)==0) {
+		infostream << "Hostname '" << hostname
+			<< "' was successfully resolved as '" 
+			<< res->ai_canonname << "'"
+			<< std::endl;
+
+		// Copy canonical name to new string so that addrinfo can be freed
```

Comment:
Yeah, this could be reworked. I started using std::string before I discovered the info|verbose|error|etc...streams. I think it would be sufficient to just use a char array. 

I noticed that there is no consensus in the code on whether to use wchar_t or char for the log streams. Are there plans to internationalize the log messages?

Reply:
No, log messages should use `char`.